### PR TITLE
WIP: DKG demo

### PR DESCRIPTION
**Pull request by conradoplg**
Creates a demo for the DKG.

It's a WIP because it's missing tests. They will be a bit complex but I'll give it a shot soon.

However I did some refactorings that may be useful for other demos:

- Use serde to read and print structs. This saves a lot of code!
- Use `impl std::io::Write` and `writeln!` for printing. This allows using the format syntax, without calling `format!` explicitly, allows removing the custom `Logger` trait, and should work for printing to terminal or to a string for testing.
- Use `Box<dyn std::error::Error>` for return types. This allows returning mostly any type of error, including FROST error or our own errors.
- Use `eyre` to return our own errors instead trying to find matching errors in the library.

### CODE REVIEW

File: dkg/src/cli.rs [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    )?;
+
+    writeln!(
+        logger,
+        "Participant key package:\n\n{}\n",
+        serde_json::to_string(&key_package)?,
+    )?;
+    writeln!(
+        logger,
+        "Partcipant public key package:\n\n{}\n",
```
Comment (natalieesk) [inline dkg/src/cli.rs:94]: typo

File: dkg/src/inputs.rs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+) -> Result<Config, Box<dyn std::error::Error>> {
+    writeln!(logger, "The minimum number of signers: (2 or more)")?;
+
+    let mut min = String::new();
+    input.read_line(&mut min)?;
+
+    let min_signers = min
+        .trim()
+        .parse::<u16>()
+        .map_err(|_| eyre!("Invalid minimum number of signers"))?;
```
Comment (natalieesk) [inline dkg/src/inputs.rs:51]: Any reason not to use the `InvalidMinSigners` error here?

File: dkg/src/inputs.rs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+    writeln!(logger, "The maximum number of signers: ")?;
+
+    let mut max = String::new();
+    input.read_line(&mut max)?;
+    let max_signers = max
+        .trim()
+        .parse::<u16>()
+        .map_err(|_| eyre!("Invalid maximum number of signers"))?;
```
Comment (natalieesk) [inline dkg/src/inputs.rs:60]: Same as above

File: dkg/src/inputs.rs [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+
+pub fn read_round2_package(
+    input: &mut impl BufRead,
+    logger: &mut dyn Write,
+) -> Result<(Identifier, round2::Package), Box<dyn std::error::Error>> {
+    writeln!(logger, "The sender's identifier (hex string):")?;
+
+    let identifier = read_identifier(input)?;
+
+    writeln!(logger, "Their JSON-encoded Round 1 Package:")?;
```
Comment (natalieesk) [inline dkg/src/inputs.rs:124]: I think this is meant to say "Their JSON-encoded Round 2 Package:"

File: dkg/src/inputs.rs [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        max_signers,
+        identifier,
+    };
+
+    validate_inputs(&config)?;
+
+    Ok(config)
+}
+
+pub fn read_identifier(input: &mut impl BufRead) -> Result<Identifier, Box<dyn std::error::Error>> {
```
Comment (natalieesk) [inline dkg/src/inputs.rs:88]: I couldn't add an entry here without getting an error. It seems the identifier_input is being requested. The identifier_input on line 103 isn't ever used. It's unclear to me what input you're trying to get here.

File: dkg/src/inputs.rs [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+    Ok(identifier)
+}
+
+pub fn read_round1_package(
+    input: &mut impl BufRead,
+    logger: &mut dyn Write,
+) -> Result<(Identifier, round1::Package), Box<dyn std::error::Error>> {
+    writeln!(logger, "The sender's identifier (hex string):")?;
+
+    let mut identifier_input = String::new();
```
Comment (natalieesk) [inline dkg/src/inputs.rs:103]: I think maybe you meant to delete lines 103 & 104? See comment above.

**Review by natalieesk — CHANGES_REQUESTED** [review]
The program is erroring at the moment. I've left a comment about why I think this is happening. Other than that just some types and questions I have about the error handling.

File: dkg/src/inputs.rs (hunk 5 above)
Comment (conradoplg) [reply dkg/src/inputs.rs:88]: I should have learned at this point to not do any refactorings right before opening the PR without proper testing :see_no_evil: Sorry about that. Fixed in 8311efe9ff96f448c8384a059f6d1cee56c8472f

File: dkg/src/inputs.rs (hunk 4 above)
Comment (conradoplg) [reply dkg/src/inputs.rs]: Exactly, fixed in 8311efe9ff96f448c8384a059f6d1cee56c8472f

File: dkg/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,16 @@
+[package]
+name = "dkg"
```
Comment (dconnolly) [inline dkg/Cargo.toml:2]: Any chance this would get published? Any better names perhaps?

Comment (natalieesk) [reply dkg/Cargo.toml:2]: We could rename them to something like frost-<project>-demo or something like that. But this won't get published so I don't think it's necessary. I'm going to merge this PR and we can discuss independently.

**Review by natalieesk — APPROVED** [review]
Looks good!

**Comment by conradoplg** [comment]
If you'd like feel free to merge this and I can work on tests in a separate PR

**Pull request merged**